hieroplatform/src/main/java/org/hiero/sketch/dataset/LocalDataSet.java

```diff
@@ -18,35 +20,46 @@ public LocalDataSet(final T data, final boolean separateThread) {
         this.separateThread = separateThread;
     }
 
+    private <R> Observable<PartialResult<R>> zero(Callable<R> z) {
+        Observable<R> zero = Observable.fromCallable(z);
+        Observable<PartialResult<R>> result = zero.map(e-> new PartialResult(0.0, e));
+        return result;
+    }
+
     @Override
     public <S> Observable<PartialResult<IDataSet<S>>> map(final IMap<T, S> mapper) {
-        final PartialResult<S> initial = new PartialResult<S>(0.0, null);
-        Observable<PartialResult<S>> start = Observable.just(initial);
+        // Callables provide lazy evaluation
+        Observable<PartialResult<IDataSet<S>>> zero = zero(() -> { return null; });
+        Callable<IDataSet<S>> r = () -> new LocalDataSet<S>(mapper.apply(this.data));
+        Observable<IDataSet<S>> start = Observable.fromCallable(r);
+        Observable<PartialResult<IDataSet<S>>> data = start.map(PartialResult::new);
+        Observable<PartialResult<IDataSet<S>>>result = zero.concatWith(data);
         if (this.separateThread)
-            start = start.observeOn(Schedulers.computation());
-        final Observable<PartialResult<S>> mapResult = mapper.apply(this.data);
-        final Observable<PartialResult<S>> chain = start.concatWith(mapResult);
-        final Observable<PartialResult<IDataSet<S>>> progress =
-                chain.map(e -> new PartialResult<IDataSet<S>>(e.deltaDone, null));
-        final PartialResultMonoid<S> monoid = new PartialResultMonoid<S>(new OptionMonoid<S>());
-        final Observable<PartialResult<S>> last = chain.reduce(monoid.zero(), monoid::add);
-        return progress.concatWith(last.map(e ->
-                new PartialResult<IDataSet<S>>(0.0, new LocalDataSet<S>(e.deltaValue))));
+            result = result.observeOn(Schedulers.computation());
+        return result;
     }
 
     @Override
     public <S> Observable<PartialResult<IDataSet<Pair<T, S>>>> zip(final IDataSet<S> other) {
         if (!(other instanceof LocalDataSet<?>))
             throw new RuntimeException("Unexpected type in Zip " + other);
-        final LocalDataSet<S> lds = (LocalDataSet<S>) other;
-        final Pair<T, S> data = new Pair<T, S>(this.data, lds.data);
-        final LocalDataSet<Pair<T, S>> retval = new LocalDataSet<Pair<T, S>>(data);
+        LocalDataSet<S> lds = (LocalDataSet<S>) other;
```

Comment:
I think we opt-in for fields, not for variables.
For variables final should be used when it clarifies the code, but I don't think we should enforce it.
It is very useful when you deal with multithreaded code and lambdas, which require effectively final variables.